src/irmin/tree_intf.ml

```diff
@@ -0,0 +1,288 @@
+      See {!force} for details about the [force] parameters. By default it is
+      [`True].
+
+      See {!uniq} for details about the [uniq] parameters. By default it is
+      [`False]. *)
+
+  (** {1 Stats} *)
+
+  type stats = {
+    nodes : int;  (** Number of node. *)
+    leafs : int;  (** Number of leafs. *)
+    skips : int;  (** Number of lazy nodes. *)
+    depth : int;  (** Maximal depth. *)
+    width : int;  (** Maximal width. *)
+  }
+  (** The type for tree stats. *)
+
+  val pp_stats : stats Fmt.t
+  (** [pp_stats] is the pretty printer for tree statistics. *)
+
+  val stats : ?force:bool -> t -> stats Lwt.t
+  (** [stats ~force t] are [t]'s statistics. If [force] is true, this will force
+      the reading of lazy nodes. By default it is [false]. *)
+
+  (** {1 Concrete Trees} *)
+
+  type concrete =
+    [ `Tree of (step * concrete) list | `Contents of contents * metadata ]
+  (** The type for concrete trees. *)
+
+  val concrete_t : concrete Type.t
+  (** The value-type for {!concrete}. *)
+
+  val of_concrete : concrete -> t
+  (** [of_concrete c] is the subtree equivalent to the concrete tree [c]. *)
+
+  val to_concrete : t -> concrete Lwt.t
+  (** [to_concrete t] is the concrete tree equivalent to the subtree [t]. *)
+
+  (** {1 Caches} *)
+
+  val clear : ?depth:int -> t -> unit
+  (** [clear ?depth t] clears all the cache in the tree [t] for subtrees with a
+      depth higher than [depth]. If [depth] is not set, all the subtrees are
+      cleared. *)
+
+  (** {1 Performance counters} *)
+
+  type counters = {
+    mutable contents_hash : int;
+    mutable contents_find : int;
+    mutable contents_add : int;
+    mutable node_hash : int;
+    mutable node_mem : int;
+    mutable node_add : int;
+    mutable node_find : int;
+    mutable node_val_v : int;
+    mutable node_val_find : int;
+    mutable node_val_list : int;
+  }
+
+  val counters : unit -> counters
+
+  val dump_counters : unit Fmt.t
+
+  val reset_counters : unit -> unit
+
+  val inspect : t -> [ `Contents | `Node of [ `Map | `Hash | `Value ] ]
+end
+
+module type Tree = sig
+  module type S = sig
+    include S
+    (** @inline *)
+  end
+
+  module Make (P : S.PRIVATE) : sig
+    include
+      S
+        with type key = P.Node.Path.t
+         and type step = P.Node.Path.step
+         and type metadata = P.Node.Val.metadata
+         and type contents = P.Contents.value
+         and type hash = P.Hash.t
+
+    val import : P.Repo.t -> P.Node.key -> node option Lwt.t
+
+    val import_no_check : P.Repo.t -> P.Node.key -> node
+
+    val export :
+      ?clear:bool ->
+      P.Repo.t ->
+      [> `Write ] P.Contents.t ->
+      [ `Read | `Write ] P.Node.t ->
+      node ->
+      P.Node.key Lwt.t
+
+    val dump : t Fmt.t
+
+    val equal : t -> t -> bool
+
+    val hash : t -> [ `Contents of hash * metadata | `Node of hash ]
+
+    val of_private_node : P.Repo.t -> P.Node.value -> node
+
+    val to_private_node : node -> P.Node.value option Lwt.t
```

Comment:
I moved the value-type into the main signature, since it seems like it belongs there. I considered moving some of these other functions, but decided not to for various reasons. e.g. `equal` is a shallow equality that doesn't force lazy nodes, so could be confusing.

Happy to be convinced otherwise.